rfc/rfc017-distributed-network-grpc-v2.md

```diff
@@ -45,4 +54,150 @@ As specified by RFC015, the node MUST authenticate the peer's node DID as follow
+
+All messages and values mentioned in this chapter are scoped to a single connection between two nodes.
+
+The protocol generally operates as follows:
+
+1. The local node broadcasts at a set interval:
+    * XOR of all known transaction references.
+    * Highest Lamport Clock value over all transactions (LC).
+    
+2. When receiving a peer's broadcast, compare the XOR value:
+    * When the XOR is the same: no action required.
+    * When different, send a message containing
+      * If the LC is incompatible with its own registration: no action required.
+      * the IBLT that includes the Lamport Clock range of 0-LC
+      * the peer LC value
+      * the own highest Lamport Clock value
+
+3. When receiving the message, the local node subtracts the given IBLT from its own IBLT for the given range:
+   * If not deconstructable, go to 1 and send values based on a smaller Lamport Clock value.
+   * If deconstructable, send a request for missing transactions
+   * If the peers highest Lamport Clock value is higher that the LC value sent in 1, then request transactions by Lamport Clock value:
+     * Lamport Clock start value.
+     * Lamport Clock end value.
+
+4. When receiving a request for transactions, respond with a message including the requested transactions.
+
+5. After adding a peer's transaction to the DAG \(making sure its cryptographic signature is valid\), query the payload if it's missing.
+
+The node MUST make sure to only add transactions of which all previous transactions are present. The node MUST also add the transaction atomically.
+If one transaction is invalid, ignore the entire message.
+
+The sections below specify the details of the protocol operation and maps it to the gRPC messages. See the section "Protobuf Definition" for a full specification of the protobuf/gRPC contract.
+
+### 4.2.1 Broadcasting
+
+The local node's DAG state MUST be broadcast at an interval using the `AdvertState` message, by default every 2 seconds. The interval MAY be adjusted by the node operator but MUST conform to the limits \(min/max interval\) defined by the network. It is advised to keep it relatively short because it directly influences the speed by which new transactions are retrieved.
+
+The `AdvertState` message contains a `XOR` value, a `LC` value and a `token`. The `token` is scoped to a connection and is required in the response from the peer.
+A node is free to choose how the `token` is created as long as it's able to satisfy the requirements of the next paragraph.
+It MUST be valid for at least 10 seconds.
+
+### 4.2.2 IBLT response
+
+When a peer receives an `AdvertState` message and the `XOR` differs, it SHOULD respond with a `TransactionSet` message.
+The upper bound of the response IBLT MUST be the equal to `512*ceil(LC/512)` where `LC` is the value from the `AdvertState` message and `ceil` is a function that rounds a double up to the nearest integer.
+This upper bound is exclusive, the lower bound is inclusive and always 0.
```

Comment:
It MUST always start at zero, since new TXs in higher ranges may depend on TXs in the lower ranges.

Other rules require ordering: did updates, vc issuance so that would prevent behaving nodes from adding everything to the root. New DIDs can all be added to the root, but given this protocol, what would be the harm? (Massive branching no longer is a problem because we're not publishing heads anymore)

Reply:
If a node gets `infected` with the collision pair, it cannot reconcile new transactions from a peer that has not been infected. This slows down propagation of new Txs throughout the network until everyone is infected.

The current protocol only describes a way to detect that the nodes are not in sync, followed by set-reconciliation. I propose we broadcast TxRefs of recently received Txs also. This could be as a separate message, or add a []TxRef to the `AdvertState`. The peer can check if it is missing any of these and request them with a `TransactionListQuery`. This should resolve the differences in most cases and reduce the need for 'heavier' set-reconciliation.